index.js

```diff
@@ -89,66 +91,107 @@ async function runEveryMinute({ global, storage, cache }) {
-
-                    if (
-                        !upcomingInvoice.error &&
-                        upcomingInvoice.created &&
-                        upcomingInvoiceDate - new Date().getTime() < ONE_DAY * global.invoiceNotificationPeriod &&
-                        upcomingInvoice.amount_due / 100 > global.invoiceAmountThreshold
-                    ) {
-                        posthog.capture('upcoming_invoice', {
-                            stripe_customer_id: customer.id,
-                            invoice_date: new Date(upcomingInvoiceDate).toLocaleDateString('en-GB'),
-                            invoice_current_amount: upcomingInvoice.amount_due / 100,
-                            distinct_id: customer.email || customer.id
-                        })
-                        await cache.set(`last_invoice_${customer.id}`, upcomingInvoiceDate)
-                    }
+                const upcomingInvoiceResponse = await fetchWithRetry(
+                    `https://api.stripe.com/v1/invoices/upcoming?customer=${customer.id}`,
+                    global.defaultHeaders
+                )
+                const upcomingInvoice = await upcomingInvoiceResponse.json()
+
+                const upcomingInvoiceDate = upcomingInvoice.created * 1000
+
+                const invoiceData = {
+                    product: productName,
+                    amount_due: upcomingInvoice.amount_due / 100,
+                    customer: customer.email,
+                    quantity: 0
+                }
+                upcomingInvoice.lines?.data.forEach((line) => {
+                    invoiceData.quantity += line.quantity
+                })
+
+                if (productName in invoicesByProduct) {
+                    invoicesByProduct[productName].push(invoiceData)
+                } else {
+                    invoicesByProduct[productName] = [invoiceData]
+                }
+
+                if (!upcomingInvoice.error && upcomingInvoice.created) {
+                    posthog.capture('Upcoming Invoice', {
+                        amount: invoiceData.amount_due,
+                        invoice_date: new Date(upcomingInvoiceDate).toLocaleDateString('en-GB'),
+                        product: invoiceData.product,
+                        quantity: invoiceData.quantity,
+                        stripe_customer_id: customer.id,
+                        distinct_id: customer.email || customer.id,
+                        $set: customer.email ? { email: customer.email } : undefined
+                    })
                 }
             }
         }
 
-        const customerRecordExists = await storage.get(customer.id)
+        // Note: I think if we want to automatically update users we should just update properties instead of firing events
+        // I also think we should perhaps restrict the properties we send so it doesn't feel overwhelming to
+        // see a lot of properties you may not understand.
```

Comment:
Yeah that was definitely a lot - the goal was to give people all the info because every company will have a different setup. I ultimately intended to either:
a) Build some schema enforcing here
b) Steer people towards using the schema enforcer plugin

Reply:
Yea, I definitely see the value in getting a lot of data in upfront and then helping make it useful downstream. I think the longer term solution here is going to be something in product (ie a taxonomy product). I like the in-between idea of only setting basic props for now, so going to go with that for now.